### PR TITLE
Minor refactoring in E2 Tokenizer, replacing if..then..elseif chains with lookups

### PR DESCRIPTION
**Pull request by texruska**
Should function the same as before, but identical & repeated ``if..then..elseif`` blocks are bad for maintainability

### CODE REVIEW

**Comment by thegrb93** [comment]
Note that in some cases ifelse is faster than a lookup table

**Comment by texruska** [comment]
I would prefer a higher quality but maringally slower codebase. Of course if this causes a major slowdown then that's a different issue

**Comment by texruska** [comment]
Rebased

**Comment by thegrb93** [comment]
If by quality, you mean readability, this is also less readable.

**Comment by CaptainPRICE** [comment]
There's a peak at which using lookup tables start to perform better than if-elseif chain, I just can't bother benchmark LuaJIT again.

And also @texruska, those lookup tables should be created statically, so, outside of the function to maximize performance, otherwise I doubt there is any performance gain with this version.

File: lua/entities/gmod_wire_expression2/base/tokenizer.lua [hunk 1]
```diff
@@ -145,44 +145,35 @@ function Tokenizer:NextSymbol()
...
+			["return"] = "ret",
+			["void"] = "void",
+			["#include"] = "inclu"
+		}
+
+		if keywords_and_tokens[self.tokendata] ~= nil then
+			tokenname = keywords_and_tokens[self.tokendata]
 		elseif self.tokendata:match("^[ijk]$") and self.character ~= "(" then
-			tokenname, self.tokendata = "num", "1" .. self.tokendata
+			tokenname = "num"
```
Comment (CaptainPRICE) [inline lua/entities/gmod_wire_expression2/base/tokenizer.lua]: This change has nothing to do with PR's intention. Why'd you split a single assignment operation into two...

Comment (texruska) [reply lua/entities/gmod_wire_expression2/base/tokenizer.lua]: I think having multiple assignments on one line is pointless just to save one LOC, plus it was more of a force of habit (Python PEP8 style). The intention of this PR is to improve readability and maintainability

**Comment by thegrb93** [comment]
```lua
local SysTime   = os.clock
local pairs     = pairs
local tostring  = tostring

local function Compare(funcs, calls)
    local results = {}
    for k, func in pairs(funcs) do
        local t = SysTime()
        for i = 1, calls do
            func()
        end
        local runtime = SysTime() - t
        results[#results+1] = {k, runtime}
    end
    table.sort(results, function(a,b) return a[2] < b[2] end)
    for k, v in pairs(results) do
        print(tostring(v[1]) .. ': ' .. v[2])
    end
end

local tokens = {
    "if",
    "elseif",
    "else",
    "local",
    "while",
    "for",
    "break",
    "continue",
    "switch",
    "case",
    "default",
    "foreach",
    "function",
    "return",
    "void",
    "#include",
}

local test_tokens = {}       
for i = 1, 1e6 do
    test_tokens[i] = tokens[math.random(#tokens)]
end

local predefined_keywords_and_tokens = {
    ["if"] = "if",
    ["elseif"] = "eif",
    ["else"] = "els",
    ["local"] = "loc",
    ["while"] = "whl",
    ["for"] = "for",
    ["break"] = "brk",
    ["continue"] = "cnt",
    ["switch"] = "swh",
    ["case"] = "case",
    ["default"] = "default",
    ["foreach"] = "fea",
    ["function"] = "func",
    ["return"] = "ret",
    ["void"] = "void",
    ["#include"] = "inclu"
}

local tokenname
Compare({
    old = function()
        
        for k, v in pairs(test_tokens) do
            if v == "if" then
                tokenname = "if"
            elseif v == "elseif" then
                tokenname = "eif"
            elseif v == "else" then
                tokenname = "els"
            elseif v == "local" then
                tokenname = "loc"
            elseif v == "while" then
                tokenname = "whl"
            elseif v == "for" then
                tokenname = "for"
            elseif v == "break" then
                tokenname = "brk"
            elseif v == "continue" then
                tokenname = "cnt"
            elseif v == "switch" then
                tokenname = "swh"
            elseif v == "case" then
                tokenname = "case"
            elseif v == "default" then
                tokenname = "default"
            elseif v == "foreach" then
                tokenname = "fea"
            elseif v == "function" then
                tokenname = "func"
            elseif v == "return" then
                tokenname = "ret"
            elseif v == "void" then
                tokenname = "void"
            elseif v == "#include" then
                tokenname = "inclu"
            end
        end
    
    end,
    new = function()
    
        for k, v in pairs(test_tokens) do
            local keywords_and_tokens = {
                ["if"] = "if",
                ["elseif"] = "eif",
                ["else"] = "els",
                ["local"] = "loc",
                ["while"] = "whl",
                ["for"] = "for",
                ["break"] = "brk",
                ["continue"] = "cnt",
                ["switch"] = "swh",
                ["case"] = "case",
                ["default"] = "default",
                ["foreach"] = "fea",
                ["function"] = "func",
                ["return"] = "ret",
                ["void"] = "void",
                ["#include"] = "inclu"
            }

            if keywords_and_tokens[v] ~= nil then
                tokenname = keywords_and_tokens[v]
            end
        end

    end,
    new_optimized = function()
    
        for k, v in pairs(test_tokens) do
            local t = predefined_keywords_and_tokens[v]
            if t ~= nil then
                tokenname = t
            end
        end

    end
}, 1)
```

Results:
new_optimized: 0.07
old: 0.12
new: 1.06


If you can implement the new_optimized version then I'd say it's acceptable.

**Comment by thegrb93** [comment]
Haven't tried with JIT yet actually. Will do that later tonight.

File: lua/entities/gmod_wire_expression2/base/tokenizer.lua [hunk 2]
```diff
@@ -145,44 +145,35 @@ function Tokenizer:NextSymbol()
...
-			tokenname = "fea"
-		elseif self.tokendata == "function" then
-			tokenname = "func"
-		elseif self.tokendata == "return" then
-			tokenname = "ret"
-		elseif self.tokendata == "void" then
-			tokenname = "void"
-		elseif self.tokendata == "#include" then
-			tokenname = "inclu"
+		local keywords_and_tokens = {
```
Comment (AbigailBuccaneer) [inline lua/entities/gmod_wire_expression2/base/tokenizer.lua]: Can this table please live in `e2lib.lua` next to `E2Lib.optable`? It'd make sense in future to be able to reuse this in the E2 text editor, plus we already have a table of operators.

Comment (CaptainPRICE) [reply lua/entities/gmod_wire_expression2/base/tokenizer.lua]: I told him in the [comment](https://github.com/wiremod/wire/pull/1916#issuecomment-489397618) already, but yeah, this should be moved outside.

File: lua/entities/gmod_wire_expression2/base/tokenizer.lua [hunk 3]
```diff
@@ -205,6 +196,17 @@ function Tokenizer:NextSymbol()
 		end
 
 	elseif self.character == "\"" then
+		-- special sequences to not escape (\n, \r, ...)
+		local special_escape_seqs = {
```
Comment (CaptainPRICE) [inline lua/entities/gmod_wire_expression2/base/tokenizer.lua]: This table as well...

**Pull request closed without merging**